File: src/main/java/jk_5/nailed/gradle/extension/NailedExtension.java
package jk_5.nailed.gradle.extension;

import com.google.common.collect.Lists;
import jk_5.nailed.gradle.delayed.DelayedString;
import jk_5.nailed.gradle.tasks.CreateLauncherProfileTask;
import lombok.Getter;
import lombok.Setter;
import org.gradle.api.Project;

import java.util.List;

/**
 * No description given
 *
 * @author jk-5
 */
public class NailedExtension {

    public static NailedExtension getInstance(Project project){
        return project.getExtensions().getByType(NailedExtension.class);
    }

    private final Project project;

    @Getter @Setter private String minecraftVersion = null;
    @Getter @Setter private String forgeVersion = null;
    @Getter @Setter private String versionName = "Nailed";
    @Getter @Setter private String loadingMavenUrl = "http://maven.reening.nl/";
    @Getter @Setter private String mainClass = null;
    @Getter @Setter private String deployHost = null;
    @Getter @Setter private String deployUsername = null;
    @Getter @Setter private String deployPassword = null;
    @Getter @Setter private String remoteProfileDir = "";
    @Getter private List<String> tweakers = Lists.newArrayList();
    @Getter private List<Project> deployed = Lists.newArrayList();
    @Getter private List<Project> deployedMods = Lists.newArrayList();
    @Getter private List<String> additionalLibs = Lists.newArrayList();

    public void setLauncherDependency(String dep){
        ((CreateLauncherProfileTask) this.project.getTasks().getByName("createLauncherProfile")).addDependency(new DelayedString(this.project, dep));
    }

    public void setAdditionalLib(String additionalLib){
        this.additionalLibs.add(additionalLib);
    }

    public void setTweaker(String tweaker){
        this.tweakers.add(tweaker);
    }

    public void setDeployedMod(String deployedMod){
        for(Project p : this.project.getSubprojects()){
            if(p.getName().equals(deployedMod)){
                this.deployedMods.add(p);
            }
        }
    }

    public void setDeployed(String deployed){
        for(Project p : this.project.getSubprojects()){
            if(p.getName().equals(deployed)){
                this.deployed.add(p);
            }
        }
    }

    public NailedExtension(Project project){
        this.project = project;
    }
}

File: src/main/scala/jk_5/nailed/gradle/extension/NailedExtension.scala
package jk_5.nailed.gradle.extension

import java.util
import org.gradle.api.Project
import jk_5.nailed.gradle.common.{UpdaterArtifact, MavenArtifact, DeployedArtifact}
import jk_5.nailed.gradle.json.RestartLevel

/**
 * No description given
 *
 * @author jk-5
 */
object NailedExtension {
  def getInstance(project: Project): NailedExtension = project.getExtensions.getByType(classOf[NailedExtension])
}

class NailedExtension(val project: Project) {

  private var minecraftVersion: String = null
  private var loadingMavenUrl: String = "http://maven.reening.nl/"
  private var deployHost: String = null
  private var deployUsername: String = null
  private var deployPassword: String = null
  private var remoteProfileDir: String = ""
  private val tweakers = new util.ArrayList[String]
  private val deployed = new util.ArrayList[UpdaterArtifact]

  @inline def getMinecraftVersion = this.minecraftVersion
  @inline def getLoadingMavenUrl = this.loadingMavenUrl
  @inline def getDeployHost = this.deployHost
  @inline def getDeployUsername = this.deployUsername
  @inline def getDeployPassword = this.deployPassword
  @inline def getRemoteProfileDir = this.remoteProfileDir
  @inline def getTweakers = this.tweakers
  @inline def getDeployed = this.deployed

  @inline def setMinecraftVersion(minecraftVersion: String) = this.minecraftVersion = minecraftVersion
  @inline def setLoadingMavenUrl(loadingMavenUrl: String) = this.loadingMavenUrl = loadingMavenUrl
  @inline def setDeployHost(deployHost: String) = this.deployHost = deployHost
  @inline def setDeployUsername(deployUsername: String) = this.deployUsername = deployUsername
  @inline def setDeployPassword(deployPassword: String) = this.deployPassword = deployPassword
  @inline def setRemoteProfileDir(remoteProfileDir: String) = this.remoteProfileDir = remoteProfileDir
  @inline def setTweaker(tweaker: String) = this.tweakers.add(tweaker)

  def setDeploy(data: util.Map[String, AnyRef]){
    if(data.containsKey("project")){
      val d = new DeployedArtifact
      d.artifact = data.get("artifact").toString
      d.projectName = data.get("project").toString
      d.mod = Option(data.get("mod")).getOrElse(false).toString.toBoolean
      d.load = Option(data.get("load")).getOrElse(false).toString.toBoolean
      d.reobf = Option(data.get("reobf")).getOrElse(false).toString.toBoolean
      d.coremod = if(data.containsKey("coremod")) data.get("coremod").toString else null
      d.restart = RestartLevel.valueOf(Option(data.get("restart")).getOrElse("nothing").toString.toUpperCase)
      d.mclib = Option(data.get("mclib")).getOrElse(false).toString.toBoolean
      this.deployed.add(d)
    }else if(data.containsKey("maven")){
      val d = new MavenArtifact
      d.artifact = data.get("artifact").toString
      d.mavenPath = data.get("maven").toString
      d.localMavenPath = if(data.containsKey("localmaven")) data.get("localmaven").toString else null
      d.mavenServer = data.get("server").toString
      d.mod = Option(data.get("mod")).getOrElse(false).toString.toBoolean
      d.load = Option(data.get("load")).getOrElse(false).toString.toBoolean
      d.reobf = Option(data.get("reobf")).getOrElse(false).toString.toBoolean
      d.coremod = if(data.containsKey("coremod")) data.get("coremod").toString else null
      d.restart = RestartLevel.valueOf(Option(data.get("restart")).getOrElse("nothing").toString.toUpperCase)
      d.mclib = Option(data.get("mclib")).getOrElse(false).toString.toBoolean
      this.deployed.add(d)
    }
  }
}
